File: Pract1/pzpi-22-10-kochevenko-ivan-design-pattern/memento.py
class RectangleMemento:
    def __init__(self, width, height, color, Xposition, Yposition, layer):
        self.width = width
        self.height = height
        self.color = color
        self.Xposition = Xposition
        self.Yposition = Yposition
        self.layer = layer

class Rectangle:
    def __init__(self, width, height, color, Xposition, Yposition, layer):
        self.width = width
        self.height = height
        self.color = color
        self.Xposition = Xposition
        self.Yposition = Yposition
        self.layer = layer

    def create_memento(self):
        return RectangleMemento(self.width, self.height, self.color, self.Xposition,
                                self.Yposition, self.layer)

    def restore_from_memento(self, memento):
        self.width = memento.width
        self.height = memento.height
        self.color = memento.color
        self.Xposition = memento.Xposition
        self.Yposition = memento.Yposition
        self.layer = memento.layer

    def display(self):
        print(f"Rectangle(width={self.width}, height={self.height}, color='{self.color}', "
              f"Xposition ='{self.Xposition}', Yposition ='{self.Yposition}', layer={self.layer})")

class History:
    def __init__(self):
        self._history = []

    def save(self, memento):
        self._history.append(memento)

    def undo(self):
        if self._history:
            return self._history.pop()
        return None

rectangle = Rectangle(100, 50, 'red', 28, 44, 1)
history = History()

rectangle.display()

history.save(rectangle.create_memento())

rectangle.width = 200
rectangle.color = 'blue'
rectangle.layer = 3
rectangle.display()

memento = history.undo()
if memento:
    rectangle.restore_from_memento(memento)

rectangle.display()

File: Pract1/pzpi-22-10-kochevenko-ivan-design-pattern/memento.cpp
#include <iostream>
#include <vector>
using namespace std;

class Rectangle {
private:
    int width, height;
    string color;
    int x, y;
    int layer;

public:
    Rectangle(int w, int h, const string& c, int x, int y, int l) : width(w), height(h), color(c), x(x), y(y), layer(l) {}

    void display() const {
        cout << "Rectangle(width=" << width
            << ", height=" << height
            << ", color=" << color
            << ", x=" << x
            << ", y=" << y
            << ", layer=" << layer << ")" << endl;
    }

    class Memento {
    private:
        int width, height;
        string color;
        int x, y;
        int layer;

        friend class Rectangle;

        Memento(int w, int h, const string& c, int x, int y, int l)
            : width(w), height(h), color(c), x(x), y(y), layer(l) {}

    public:
        Memento(const Memento& other) = default;
        ~Memento() = default;
    };

    Memento save() const {
        return Memento(width, height, color, x, y, layer);
    }

    void restore(const Memento& memento) {
        width = memento.width;
        height = memento.height;
        color = memento.color;
        x = memento.x;
        y = memento.y;
        layer = memento.layer;
    }
};

class History {
private:
    vector<Rectangle::Memento> history;

public:
    void save(const Rectangle::Memento& memento) {
        history.push_back(memento);
    }

    Rectangle::Memento undo() {
        if (!history.empty()) {
            Rectangle::Memento m = history.back();
            history.pop_back();
            return m;
        }
        throw runtime_error("History is empty!");
    }
};

int main() {
    Rectangle rect(100, 50, "red", 50, 40, 1);
    History history;

    rect.display();
    history.save(rect.save());

    rect = Rectangle(200, 100, "blue", 88, 41, 2);
    rect.display();

    rect.restore(history.undo());
    rect.display();

    return 0;
}
